src/librender/integrator.cpp

```diff
@@ -312,11 +319,190 @@ MTS_VARIANT MonteCarloIntegrator<Float, Spectrum>::MonteCarloIntegrator(const Pr
+        Log(Warn, "Scene does not contain any emitter, returning black image.");
+        normalize_film(film, 1);
+        return true;
+    }
+
+    ThreadEnvironment env;
+    tbb::spin_mutex mutex;
+    ref<ProgressReporter> progress = new ProgressReporter("Rendering");
+
+    size_t n_cores = util::core_count();
+    /* Multiply sample count by pixel count to obtain a similar scale
+     * to the standard path tracer. */
+    size_t total_samples =
+        sensor->sampler()->sample_count() * hprod(film->size());
+    // TODO: handle different modes
+    size_t iteration_count = total_samples;
+    // size_t iteration_count = IsVectorized
+    //                              ? std::rint(total_samples / (Float) PacketSize)
+    //                              : total_samples;
+    size_t grain_count = std::max(
+        (size_t) 1, (size_t) std::rint(iteration_count / Float(n_cores * 2)));
+
+    // Insert default channels and set up the film
+    std::vector<std::string> channels = { "X", "Y", "Z", "A", "W" };
+    film->prepare(channels);
+    bool has_aovs = false;
+    if (has_aovs)
+        Throw("Not supported yet: AOVs in LightTracerIntegrator");
+
+    Log(Info, "Starting render job (%ix%i, %i sample%s, %i core%s)",
+        film->crop_size().x(), film->crop_size().y(), total_samples,
+        total_samples == 1 ? "" : "s", n_cores, n_cores == 1 ? "" : "s");
+    if (m_timeout > 0.0f)
+        Log(Info, "Timeout specified: %.2f seconds.", m_timeout);
+
+    size_t total_samples_done = 0;
+    tbb::parallel_for(
+        tbb::blocked_range<size_t>(0, iteration_count, grain_count),
+        [&](const tbb::blocked_range<size_t> &range) {
+            ScopedSetThreadEnvironment set_env(env);
+
+            // Thread-specific accumulators.
+            // Enable warning, border and normalization for the image block.
+            ref<ImageBlock> block = new ImageBlock(
+                film->size(), channels.size(), film->reconstruction_filter(),
+                /* warn_negative */ !has_aovs,
+                /* warn_invalid */ true, /* border */ true,
+                /* normalize */ true);
+            block->clear();
+            size_t samples_done = 0;
+
+            // Ensure that sample generation is fully deterministic.
+            ref<Sampler> sampler = sensor->sampler()->clone();
+            // TODO: appropriate seeding in wavefront mode
+            size_t seed = (size_t) range.begin();
+            sampler->seed(seed);
+
+            for (auto i = range.begin(); i != range.end() && !should_stop(); ++i) {
+                if (likely(m_max_depth != 0)) {
+                    // Account for visible emitters
+                    sample_visible_emitters(scene, sensor, sampler, block);
+                }
+
+                // Primary & further bounces illumination
+                auto [ray, throughput] = prepare_ray(scene, sensor, sampler);
+
+                // TODO(!): how to create a correct initial si?
```

Comment:
For scalar variants this initialization should be okay. Otherwise we usually use `ek::zero<SurfaceInteraction3f>()` and then set `si.t = ek::Infinity<Float>` to ensure the surface interaction is invalid.